File: NeetCode/Backtracking/Permutations_II/permutations__i_i.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        results = []
        used = [False] * len(nums)

        def dfs(curr: List[int]) -> None:
            if len(curr) == len(nums):
                results.append(curr.copy())
                return

            for i in range(len(nums)):
                if used[i]:
                    continue

                # Only use the first unused copy of a duplicate value.
                if i > 0 and nums[i] == nums[i - 1] and not used[i - 1]:
                    continue

                used[i] = True
                curr.append(nums[i])

                dfs(curr)

                curr.pop()
                used[i] = False

        dfs([])
        return results


if __name__ == "__main__":
    sol = Solution()
    nums = [1, 1, 2]
    print(sol.permuteUnique(nums))
    print("Running Solution...")


""" 
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        results = set()
        n = len(nums)

        if n == 1:
            return [nums]

        for index in range(n):

            arr = nums.copy()
            tmp = arr.pop(index)

            permutations = self.permuteUnique(arr)

            for perm in permutations:
                t = [tmp] + perm
                results.add(tuple(t))

        return [[i for i in perm] for perm in results]

"""

File: NeetCode/Backtracking/Permutations_II/Permutations_II.cpp
// Permutations_II.cpp
//
// Compile locally with:
// clang++ Permutations_II.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<vector<int>> permuteUnique(vector<int> &nums)
    {
        sort(nums.begin(), nums.end());

        vector<vector<int>> results;
        vector<bool> used(nums.size(), false);
        vector<int> curr;
        dfs(curr, nums, used, results);
        return results;
    }

    void dfs(vector<int> &curr, vector<int> &nums, vector<bool> &used, vector<vector<int>> &results)
    {
        if (curr.size() == nums.size())
        {
            results.push_back(curr);
            return;
        }

        for (int index = 0; index < nums.size(); index++)
        {
            if (used[index])
            {
                continue;
            }

            if (index > 0 && nums[index] == nums[index - 1] and !used[index - 1])
            {
                continue;
            }

            used[index] = true;
            curr.push_back(nums[index]);
            dfs(curr, nums, used, results);
            curr.pop_back();
            used[index] = false;
        }
    }
};

int main()
{
    cout << "Running Permutations_II.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
